# Geometry optimization

We use DFTK and the [GeometryOptimization](https://github.com/JuliaMolSim/GeometryOptimization.jl/)
package to find the minimal-energy bond length of the $H_2$ molecule.
First we set up an appropriate `DFTKCalculator` (see AtomsCalculators integration),
for which we use the LDA model just like in the Tutorial for silicon.

In [1]:
using DFTK

calc = DFTKCalculator(;
    model_kwargs = (; functionals=LDA()),        # model_DFT keyword arguments
    basis_kwargs = (; kgrid=[1, 1, 1], Ecut=10)  # PlaneWaveBasis keyword arguments
)

DFTKCalculator{Nothing}(DFTK.DFTKParameters((functionals = Xc(lda_x, lda_c_pw),), (kgrid = [1, 1, 1], Ecut = 10), (callback = identity,)), nothing, Base.RefValue{Int64}(0), Base.RefValue{Int64}(0), true)

Next we set up an initial hydrogen molecule within a box of vacuum.
We use the parameters of the
[equivalent tutorial from ABINIT](https://docs.abinit.org/tutorial/base1/),
that is a simulation box of 10 bohr times 10 bohr times 10 bohr and a
pseudodojo pseudopotential.

In [2]:
using LinearAlgebra
using LazyArtifacts
import Main: @artifact_str  # hide

r0 = 1.4   # Initial bond length in Bohr
a  = 10.0  # Box size in Bohr

lattice = a * I(3)
H = ElementPsp(:H; psp=load_psp(artifact"pd_nc_sr_pbe_standard_0.4.1_upf/H.upf"));
atoms = [H, H]
positions = [zeros(3), lattice \ [r0, 0., 0.]]

h2_crude = periodic_system(lattice, atoms, positions)

FlexibleSystem(H₂, periodic = TTT):
    bounding_box      : [      10        0        0;
                                0       10        0;
                                0        0       10]u"a₀"

    Atom(H,  [       0,        0,        0]u"a₀")
    Atom(H,  [     1.4,        0,        0]u"a₀")

       .------------.  
      /|            |  
     / |            |  
    *  |            |  
    |  |            |  
    |  |            |  
    |  .------------.  
    | /            /   
    |/            /    
    H-H----------*     


Finally we call `minimize_energy!` to start the geometry optimisation.
We use `verbosity=2` to get some insight into the minimisation.
With `verbosity=1` only a summarising table would be printed and with
`verbosity=0` (default) the minimisation would be quiet.

In [3]:
using GeometryOptimization
results = minimize_energy!(h2_crude, calc; tol_forces=2e-6, verbosity=2)
nothing  # hide

n     Energy            log10(ΔE)   log10(Δρ)   α      Diag   Δtime
---   ---------------   ---------   ---------   ----   ----   ------
  1   -1.110606920628                   -0.82   0.80    7.0    4.12s
  2   -1.117203404512       -2.18       -1.86   0.80    1.0    1.85s
  3   -1.117251870239       -4.31       -2.71   0.80    1.0   17.2ms
  4   -1.117252481144       -6.21       -3.56   0.80    1.0   36.4ms
  5   -1.117252499159       -7.74       -4.04   0.80    1.0   18.4ms
  6   -1.117252500752       -8.80       -4.81   0.80    1.0   17.9ms
  7   -1.117252500843      -10.04       -5.57   0.80    1.0   20.4ms
  8   -1.117252500846      -11.52       -6.27   0.80    2.0   23.1ms
  9   -1.117252500846      -12.97       -6.67   0.80    1.0   20.6ms
 10   -1.117252500846      -13.82       -8.03   0.80    1.0   19.2ms
 11   -1.117252500846   +  -15.05       -8.12   0.80    2.0   64.3ms
 12   -1.117252500846   +  -14.75       -8.79   0.80    1.0    366ms
n     Energy            log10(ΔE)  

Structure after optimisation (note that the atom has wrapped around)

In [4]:
results.system

FlexibleSystem(H₂, periodic = TTT):
    bounding_box      : [      10        0        0;
                                0       10        0;
                                0        0       10]u"a₀"

    Atom(H,  [-0.0431826, -1.57733e-11, 5.10406e-12]u"a₀")
    Atom(H,  [ 1.44318, -1.74866e-11, 1.93959e-11]u"a₀")

       .------------.  
      /|            |  
     / |            |  
    *  |            |  
    |  |            |  
    |  |            |  
    |  .-H----------H  
    | /            /   
    |/            /    
    *------------*     


Compute final bond length:

In [5]:
rmin = norm(position(results.system[1]) - position(results.system[2]))
println("Optimal bond length: ", rmin)

Optimal bond length: 1.4863652663708993 a₀


Our results (1.486 Bohr) agrees with the
[equivalent tutorial from ABINIT](https://docs.abinit.org/tutorial/base1/).